rust/parser/mod.rs

```diff
@@ -20,5 +20,601 @@
  *
  */
 
-pub mod parser;
-pub mod syntax_error;
\ No newline at end of file
+pub mod error_listener;
+pub mod syntax_error;
+
+#[cfg(test)]
+mod test;
+
+use antlr_rust::token::Token;
+use antlr_rust::tree::ParseTree;
+use antlr_rust::tree::TerminalNode as ANTLRTerminalNode;
+use chrono::{NaiveDateTime, Timelike};
+use std::rc::Rc;
+
+use crate::common::error::{ErrorMessage, ILLEGAL_GRAMMAR, ILLEGAL_STATE};
+use typeql_grammar::typeqlrustparser::*;
+
+use crate::pattern::*;
+use crate::query::*;
+use crate::typeql_match;
+
+#[derive(Debug)]
+pub struct Definable;
+
+type ParserResult<T> = Result<T, ErrorMessage>;
+type TerminalNode<'a> = ANTLRTerminalNode<'a, TypeQLRustParserContextType>;
+
+#[derive(Debug)]
+enum Type {
+    Unscoped(String),
+    Scoped(ScopedType),
+    Variable(TypeVariable),
+}
+
+fn get_string(string: Rc<TerminalNode>) -> String {
+    let quoted = string.get_text();
+    String::from(&quoted[1..quoted.len() - 1])
+}
+
+fn get_long(long: Rc<TerminalNode>) -> ParserResult<i64> {
+    long.get_text()
+        .parse()
+        .map_err(|_| ILLEGAL_GRAMMAR.format(&[long.get_text().as_str()]))
+}
+
+fn get_double(double: Rc<TerminalNode>) -> ParserResult<f64> {
+    double
+        .get_text()
+        .parse()
+        .map_err(|_| ILLEGAL_GRAMMAR.format(&[double.get_text().as_str()]))
+}
+
+fn get_date(date: Rc<TerminalNode>) -> ParserResult<NaiveDateTime> {
+    NaiveDateTime::parse_from_str(&date.get_text(), "%Y-%m-%d")
+        .map_err(|_| ILLEGAL_GRAMMAR.format(&[date.get_text().as_str()]))
+}
+
+fn parse_date_time(date_time_text: &str) -> Option<NaiveDateTime> {
+    let has_seconds = date_time_text.matches(":").count() == 2;
+    if has_seconds {
+        let has_nanos = date_time_text.matches(".").count() == 1;
+        if has_nanos {
+            let parts: Vec<&str> = date_time_text.splitn(2, ".").collect();
+            let (date_time, nanos) = (parts[0], parts[1]);
+            NaiveDateTime::parse_from_str(date_time, "%Y-%m-%dT%H:%M:%S")
+                .ok()?
+                .with_nanosecond(
+                    format!("{}{}", nanos, "0".repeat(9 - nanos.len()))
+                        .parse()
+                        .ok()?,
+                )
+        } else {
+            NaiveDateTime::parse_from_str(date_time_text, "%Y-%m-%dT%H:%M:%S").ok()
+        }
+    } else {
+        NaiveDateTime::parse_from_str(date_time_text, "%Y-%m-%dT%H:%M").ok()
+    }
+}
```

Comment:
It's not even really modifying the grammar, just telling the parser to keep the datetime tokens.

It could live alongside the custom datetime formatter, I suppose.